crates/router/src/connector/rapyd.rs

```diff
@@ -642,25 +693,145 @@ impl services::ConnectorIntegration<api::RSync, types::RefundsData, types::Refun
+            .await
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+        let message = self
+            .get_webhook_source_verification_message(headers, body, merchant_id, &secret)
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+
+        let stringify_auth = String::from_utf8(secret.to_vec())
+            .into_report()
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)
+            .attach_printable("Could not convert secret to UTF-8")?;
+        let auth: transformers::RapydAuthType = stringify_auth
+            .parse_struct("RapydAuthType")
+            .change_context(errors::ConnectorError::WebhookSourceVerificationFailed)?;
+        let secret_key = auth.secret_key;
+        let key = hmac::Key::new(hmac::HMAC_SHA256, secret_key.as_bytes());
+        let tag = hmac::sign(&key, &message);
+        let hmac_sign = hex::encode(tag);
+        Ok(hmac_sign.as_bytes().eq(&signature))
+    }
+
+    fn get_webhook_object_reference_id(
+        &self,
+        body: &[u8],
     ) -> CustomResult<String, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let webhook: transformers::RapydIncomingWebhook = body
+            .parse_struct("RapydIncomingWebhook")
+            .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
+        Ok(match webhook.data {
+            transformers::WebhookData::PaymentData(payment_data) => payment_data.id,
+            transformers::WebhookData::RefundData(refund_data) => refund_data.id,
+        })
     }
 
     fn get_webhook_event_type(
         &self,
-        _body: &[u8],
+        body: &[u8],
     ) -> CustomResult<api::IncomingWebhookEvent, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let webhook: transformers::RapydIncomingWebhook = body
+            .parse_struct("RapydIncomingWebhook")
+            .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
+        webhook.webhook_type.try_into()
     }
 
     fn get_webhook_resource_object(
         &self,
-        _body: &[u8],
+        body: &[u8],
     ) -> CustomResult<serde_json::Value, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let webhook: transformers::RapydIncomingWebhook = body
+            .parse_struct("RapydIncomingWebhook")
+            .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
+        let response = match webhook.data {
+            transformers::WebhookData::PaymentData(payment_data) => {
+                let rapyd_response: transformers::RapydPaymentsResponse = payment_data.into();
+                Ok(rapyd_response)
+            }
+            _ => Err(errors::ConnectorError::WebhookEventTypeNotFound),
+        }?;
+        let res_json = serde_json::to_value(response)
```

Comment:
check in common_utils crate, ext_traits.rs
check any common utility function which does the same.

Reply:
okay